cloud/pkg/devicecontroller/controller/downstream.go

```diff
@@ -325,6 +389,74 @@ func (dc *DownstreamController) sendDeviceModelMsg(device *v1beta1.Device, opera
 	}
 }
 
+// mapperDeleted delete records about node with nodeID in mapperManager
+func (dc *DownstreamController) mapperMigrated(nodeID string) error {
+	// When the node is disconnected, the mapper on the node needs to be deleted
+	value, ok := dc.mapperManager.NodeMapperList.Load(nodeID)
+	if ok {
+		mapperList := value.([]*pb.MapperInfo)
+		for _, mapper := range mapperList {
+			// delete mapper in Mapper2NodeMap
+			dc.mapperManager.Mapper2NodeMap.Delete(mapper.Name)
+			msg := model.NewMessage("")
+			resource, err := messagelayer.BuildResourceForDevice(nodeID, "mapper", "")
+			if err != nil {
+				klog.Errorf("Built message resource failed with error: %s", err)
+			}
+			msg.BuildRouter("meta", constants.GroupTwin, resource, model.DeleteOperation)
+			msg.Content, err = json.Marshal(mapper)
+			if err != nil {
+				klog.Warningf("Failed to marshal mapper %v due to error %v", mapper, err)
+			}
+			err = dc.messageLayer.Send(*msg)
+			if err != nil {
+				klog.Errorf("Failed to send mapper deletion message %v due to error %v", msg, err)
+			}
+		}
+		// delete nodeId in NodeMapperList
+		dc.mapperManager.NodeMapperList.Delete(nodeID)
+	}
+	return nil
+}
+
+// deviceMigrated indicates the device to be migrated on the node with nodeID
+func (dc *DownstreamController) deviceMigrated(nodeID string) error {
+	value, ok := dc.deviceManager.NodeDeviceList.Load(nodeID)
+	if ok {
+		deviceList := value.([]*v1beta1.Device)
+		for i := range deviceList {
+			// if device.Spec.MigrateOnOffline is true, enable device migration
+			if deviceList[i].Spec.MigrateOnOffline {
+				device := deviceList[i].DeepCopy()
+				device.Status.CurrentNode = ""
+				device.Spec.NodeName = ""
+				dc.deviceUpdated(device)
+				i--
+			}
+		}
```

Comment:
How about deepcopy a new devicelist here instead of using i--? It's easier to understand